File: Problems 0000-0099/04. Median of Two Sorted Arrays/solution.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int m = nums1.size(), n = nums2.size();
        if (m > n) {
            return findMedianSortedArrays(nums2, nums1);
        }

        // num from nums1
        int lo = 0, hi = m;
        int numOnLeftSide = (m + n + 1) / 2;
        while (lo <= hi) {
            int mid1 = lo + (hi - lo) / 2;
            int mid2 = numOnLeftSide - mid1;

            int l1 = mid1 > 0 ? nums1[mid1-1] : INT_MIN, r1 = mid1 < m ? nums1[mid1] : INT_MAX;
            int l2 = mid2 > 0 ? nums2[mid2-1] : INT_MIN, r2 = mid2 < n ? nums2[mid2] : INT_MAX;
            if (l1 <= r2 && l2 <= r1) {
                // correct partition!
                if ((m + n) % 2 == 0) {
                    return (double)(max(l1, l2) + min(r1, r2)) / 2.0;
                }
                return max(l1, l2);
            } else if (l1 > r2) {
                hi = mid1 - 1;
            } else {
                lo = mid1 + 1;
            }

        }
        return -1;
    }
};

File: Problems 0000-0099/04. Median of Two Sorted Arrays/solution.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        m, n = len(nums1), len(nums2)

        if m > n:
            return self.findMedianSortedArrays(nums2, nums1)

        targetNum = (m + n + 1) // 2
        low = 0
        high = m

        # Finding not on index, but searching partition size
        while low <= high:
            numOnLeftSide1 = low + (high - low) // 2
            numOnLeftSide2 = targetNum - numOnLeftSide1

            l1 = float('-inf')
            l2 = float('-inf')
            r1 = float('inf')
            r2 = float('inf')
            if numOnLeftSide1 > 0:
                l1 = nums1[numOnLeftSide1 - 1]
            if numOnLeftSide2 > 0:
                l2 = nums2[numOnLeftSide2 - 1]
            if numOnLeftSide1 < m:
                r1 = nums1[numOnLeftSide1]
            if numOnLeftSide2 < n:
                r2 = nums2[numOnLeftSide2]

            if l1 <= r2 and l2 <= r1:
                if (m + n) % 2 == 0:
                    return (max(l1, l2) + min(r1, r2)) / 2.0
                else:
                    return max(l1, l2)
            elif l1 > r2:
                high = numOnLeftSide1 - 1
            else:
                low = numOnLeftSide1 + 1
